File: solutions/r/queen-attack/1/queen-attack.R
create <- function(row, col) {
  if (row < 0 || row > 7) {
    stop("row must be between 0 and 7")
  }

  if (col < 0 || col > 7) {
    stop("column must be between 0 and 7")
  }

  c(row = row, col = col)
}

can_attack <- function(queen1, queen2) {
  same_row <- queen1["row"] == queen2["row"]
  same_col <- queen1["col"] == queen2["col"]

  same_diagonal <- abs(queen1["row"] - queen2["row"]) ==
                   abs(queen1["col"] - queen2["col"])

  same_row || same_col || same_diagonal
}

File: solutions/delphi/queen-attack/1/queen_attack.pas
unit uQueenAttack;

interface

type
  TQueen = class
  private
    x,y:integer;
  public
    function CanAttack(aQuenn:TQueen):boolean;
    constructor Create(const aRow, aCol:integer);
  end;

implementation

uses
  System.SysUtils;

{ TQueen }

function TQueen.CanAttack(aQuenn: TQueen): boolean;
begin
  Result := (x = aQuenn.x) or (y = aQuenn.y) or           // same row,col
            ((x + y) = (aQuenn.x + aQuenn.y)) or          // first + second diagonal
            (Abs(x - y) = Abs(aQuenn.x - aQuenn.y));      // third + four diagonal
end;

constructor TQueen.Create(const aRow, aCol: integer);
begin
  inherited Create;

  if aRow < 0 then
    raise EArgumentException.Create('row not positive');
  if acol < 0 then
    raise EArgumentException.Create('column not positive');
  if (acol > 7)  then
    raise EArgumentException.Create('column not on board');
  if (aRow > 7) then
    raise EArgumentException.Create('row not on board');

  x := aRow;
  y := aCol;
end;

end.